keps/sig-instrumentation/4785-crdmetrics-controller/README.md

```diff
@@ -0,0 +1,1061 @@
+* Garbage in, garbage out: Invalid configurations will generate invalid metrics.
+  The exception to this being that certain checks that ensure metric structure
+  are still present (for e.g., `value` should be a `float64`).
+* Library support: The module is **not** intended to be used as a library, and 
+  as such, does not export any functions or types, with `pkg/` being an 
+  exception (for e.g., managed resource types).
+* Metrics stability: There are **no** metrics stability guarantees, as the
+  metrics are dynamically generated.
+* No middle-ware: The configuration is `unmarshal`led into a set of stores that
+  the codebase directly operates on. Unlike Kube State Metrics, there is **no**
+  controller-defined parsing middle-ware that processes the configuration before
+  it is used, in order to cut down on unnecessary complexity as much as
+  possible.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+N/A since the managed resource offered by the CRDMetrics controller provides the
+ability to define metric configurations that are a super-set of the 
+expressibility that Kube State Metrics' Custom Resource State configurations
+have to offer.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The controller offers a number of improvements over Kube State Metrics' Custom
+Resource State API, while maintaining a 3x faster round trip time for metric
+generation.
+
+- At its core, the controller relies on its managed resource,
+`CRDMetricsResource` to fetch the metric generation configuration. Parts of the
+configuration may be defined using different `resolver`s, such as `unstructured`
+or `CEL`.
+- Once fetched, the controller `unmarshal`s the configuration YAML directly into
+`stores` which are a set of metric `families`, which in turn are a set of 
+`metrics`.
+- Metric `stores` are created based on its respective GVKR (a type that embeds 
+`schema.GroupVersionKind`, `schema.GroupVersionResource` to avoid 
+[plural ambiguities]), and reflectors for the specified resource are
+initialized, and populate the stores on its update.
+- `/metrics` pings on `CRDMETRICS_MAIN_PORT` trigger the server to write the
+raw metrics, combined with its appropriate header(s), in the response. All
+generated metrics are hardcoded to `gauge`s by design, as Prometheus lacks
+support for some OpenMetrics-specified metrics' types, such as `Info` and
```

Comment:
I learned recently that OpenMetrics has folded into Prometheus and is no longer A Thing.

Reply:
Yep, we'd like to persist this information for anyone coming from `kube-state-metrics`, which still uses the OpenMetrics types stated above, for the sake of relevance.